plugins/unreal/Open3DStream/Source/Open3DStream/Private/Open3DServer.cpp

```diff
@@ -1,39 +1,54 @@
 #include "Open3DServer.h"
+#include "o3ds/async_pair.h"
 
+void InDataFunc(void *ptr, void *data, size_t msg)
+{
+	static_cast<UServer*>(ptr)->inData((const char*)data, msg);
+}
 
-USubscriber::USubscriber()
+UServer::UServer()
+	: mServer(nullptr)
 {}
 
-void USubscriber::in_data(const char *msg, size_t len)
+bool UServer::start(const char *url, const char *protocol)
 {
-	mutex.Lock();
-	buffer.resize(len);
-	buffer.assign(msg, msg + len);
-	mutex.Unlock();
+	if (mServer)
+	{
+		delete(mServer);
+		mServer = nullptr;
+			
+	}
+	if (strcmp(protocol, "Pair-Client") == 0)
+	{
+		mServer = new O3DS::AsyncPair();
+		mServer->setFunc(this, InDataFunc);
+		mServer->connect(url);
+		return true;
+	}
+	if (strcmp(protocol, "Pair-Server") == 0)
+	{
+		mServer = new O3DS::AsyncPair();
+		mServer->setFunc(this, InDataFunc);
+		mServer->listen(url);
+		return true;
+	}
 
-	DataDelegate.ExecuteIfBound();
+	return false;
 }
 
-void USubscriber::in_pipe()
-{}
-
-/*
-bool USubscriber::Recv()
+bool UServer::write(const char *msg, size_t len)
 {
-	nng_msg *msg;
-	int ret;
-	if ((ret = nng_recvmsg(mSocket, &msg, 0)) != 0) {
-		return false;
-	}
+	if (!mServer) return false;
```

Comment:
Instead of returning false when the server is null why not throw an exception? This way a user of the API is forced to do something about it.